### PR TITLE
Add support to VB for custom modifiers associated with type arguments of generic types.

### LINKED ISSUE #4163: VB & C# cannot consume method with modopt
I have a C++ library that defines a member like
```C++/CLI
property Nullable<DWORD> Type{Nullable<DWORD> get();}
```
which is translated to CLI as
```CIL
.property instance valuetype [mscorlib]System.Nullable`1<uint32 modopt([mscorlib]System.Runtime.CompilerServices.IsLong)> Type
{
    .get instance valuetype [mscorlib]System.Nullable`1<uint32 modopt([mscorlib]System.Runtime.CompilerServices.IsLong)> Dzonny.XmlKeyboard.Interop.KbdTables::get_Type()
}
```

VB complains that it cannot consume it because "Property 'Type' is of an unsupported type."

This used to work before Roslyn, and it should work according to CTS specification consumer must read and correctly propagate all modopts, even if it does not understand them.

C# has the same issue, but with more misleading error message:  "Property, indexer, or event 'KbdTables.Type' is not supported by the language; try directly calling accessor method 'KbdTables.get_Type()'." But when I try to call get_Type() directly I get: "'KbdTables.get_Type()' is not supported by the language."

### PR DESCRIPTION
**Pull request by AlekseyTs**
Fixes #4163.

@gafter, @jaredpar, @VSadov, @agocke Please review.

### CODE REVIEW

**Comment by AlekseyTs** [comment]
@dotnet-bot Test this please

File: src/Compilers/CSharp/Portable/Emitter/Model/ParameterSymbolAdapter.cs [hunk 1]
```diff
@@ -29,11 +29,11 @@ bool Cci.IParameterTypeInformation.IsByReference
             }
         }
 
-        bool Cci.IParameterTypeInformation.HasByRefBeforeCustomModifiers
+        ushort Cci.IParameterTypeInformation.CountOfCustomModifiersPrecedingByRef
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Emitter/Model/ParameterSymbolAdapter.cs]: Is there a limit on the number of custom modifiers that can appear which caused us to choose `ushort` here?

File: src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEParameterSymbol.cs [hunk 2]
```diff
@@ -232,10 +223,68 @@ public bool TryGetWellKnownAttribute(WellKnownAttributeFlags flag, out bool valu
...
+                ushort countOfCustomModifiersPrecedingByRef,
+                TypeSymbol type,
+                ParameterHandle handle,
+                ImmutableArray<ModifierInfo<TypeSymbol>> customModifiers,
+                out bool isBad) :
+                    base(moduleSymbol, containingSymbol, ordinal, isByRef, type, handle, customModifiers.Length, out isBad)
+            {
+                _customModifiers = CSharpCustomModifier.Convert(customModifiers);
+                _countOfCustomModifiersPrecedingByRef = countOfCustomModifiersPrecedingByRef;
+
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEParameterSymbol.cs]: Would it make sense to assert here that `_countOfCustomModifiersPrecedingRef <= customModifiers.Length`?

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceComplexParameterSymbol.cs [hunk 3]
```diff
@@ -1047,4 +1040,49 @@ internal override void ForceComplete(SourceLocation locationOpt, CancellationTok
             var unused = this.ExplicitDefaultConstantValue;
         }
     }
+
+    internal sealed class SourceComplexParameterSymbolWithCustomModifiers : SourceComplexParameterSymbol
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceComplexParameterSymbol.cs]: I didn't think we could define parameters with customer modifiers in source.  Can you give an example of when this type would be used?

File: src/Compilers/CSharp/Portable/Emitter/Model/ParameterSymbolAdapter.cs (hunk 1 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Emitter/Model/ParameterSymbolAdapter.cs]: I do not remember seen any specified limit, but I believe it is reasonable for us to not support signatures with amount of custom modifiers that doesn't fit into ushort. We can even go to byte, but one might argue that having more than 255 modifiers is something that could happen in the wild.

File: src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEParameterSymbol.cs (hunk 2 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEParameterSymbol.cs]: Sure, an extra assert here wouldn't hurt.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceComplexParameterSymbol.cs (hunk 3 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceComplexParameterSymbol.cs]: This happens when we override a method that has modifiers. The modifiers are inherited from the overriden method.

**Comment by VSadov** [comment]
LGTM
It is going to be "fun" to merge with ref returns...

File: src/Compilers/CSharp/Portable/Symbols/MemberSignatureComparer.cs [hunk 4]
```diff
@@ -602,7 +602,7 @@ private static bool HaveSameParameterTypes(ImmutableArray<ParameterSymbol> param
                 {
                     if (!type1.Equals(type2, ignoreDynamic: true) ||
                         !param1.CustomModifiers.SequenceEqual(param2.CustomModifiers) ||
-                        (!param1.CustomModifiers.IsEmpty && param1.RefKind != RefKind.None && param2.RefKind != RefKind.None && param1.HasByRefBeforeCustomModifiers != param2.HasByRefBeforeCustomModifiers))
+                        (param1.RefKind != RefKind.None && param2.RefKind != RefKind.None && param1.CountOfCustomModifiersPrecedingByRef != param2.CountOfCustomModifiersPrecedingByRef))
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Symbols/MemberSignatureComparer.cs]: Could you omit the test `param1.RefKind != RefKind.None && param2.RefKind != RefKind.None &&` here?

File: src/Compilers/CSharp/Portable/Symbols/Metadata/PE/SymbolFactory.cs [hunk 5]
```diff
@@ -153,7 +154,12 @@ internal override TypeSymbol GetUnsupportedMetadataTypeSymbol(PEModuleSymbol mod
                 return new UnsupportedMetadataTypeSymbol();
             }
 
-            TypeMap substitution = new TypeMap(typeParameters, arguments);
+            if (System.Linq.Enumerable.Any(arguments, arg => !arg.Value.IsDefaultOrEmpty))
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Symbols/Metadata/PE/SymbolFactory.cs]: Can you please make this a plain old loop?

File: src/Compilers/CSharp/Portable/Symbols/MemberSignatureComparer.cs (hunk 4 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/MemberSignatureComparer.cs]: I'll investigate if it makes sense to drop this condition too.

File: src/Compilers/CSharp/Portable/Symbols/Metadata/PE/SymbolFactory.cs (hunk 5 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Metadata/PE/SymbolFactory.cs]: This code is going to be replaced with a real handling of custom modifiers in the next change.

File: src/Compilers/VisualBasic/Portable/Lowering/AsyncRewriter/AsyncRewriter.AsyncMethodToClassRewriter.Await.vb [hunk 6]
```diff
@@ -28,7 +28,7 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
                 ' The awaiter temp facilitates EnC method remapping and thus have to be long-lived.
                 ' It transfers the awaiter objects from the old version of the MoveNext method to the New one.
                 Debug.Assert(node.Syntax.IsKind(SyntaxKind.AwaitExpression))
-                Dim awaiterType As TypeSymbol = node.GetAwaiter.Type.InternalSubstituteTypeParameters(Me.TypeMap)
+                Dim awaiterType As TypeSymbol = node.GetAwaiter.Type.InternalSubstituteTypeParameters(Me.TypeMap).AsTypeSymbol()
```
Comment (gafter) [inline src/Compilers/VisualBasic/Portable/Lowering/AsyncRewriter/AsyncRewriter.AsyncMethodToClassRewriter.Await.vb]: The assertion in AsTypeSymbol could fail if the library's GetAwaiter() method has a custom modifier. Is an assertion failure the behavior we want?

File: src/Compilers/VisualBasic/Portable/Symbols/NamedTypeSymbol.vb [hunk 7]
```diff
@@ -48,6 +48,23 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
...
+        ''' <summary>
+        ''' Returns custom modifiers for the type arguments that have been substituted for the type parameters. 
+        ''' </summary>
+        Friend MustOverride ReadOnly Property TypeArgumentsCustomModifiers As ImmutableArray(Of ImmutableArray(Of CustomModifier))
+
+        Friend Function CreateEmptyTypeArgumentsCustomModifiers() As ImmutableArray(Of ImmutableArray(Of CustomModifier))
+            Dim arity = Me.Arity
+
+            If arity > 0 Then
+                Return ArrayBuilder(Of ImmutableArray(Of CustomModifier)).GetInstance(arity, ImmutableArray(Of CustomModifier).Empty).ToImmutableAndFree()
```
Comment (gafter) [inline src/Compilers/VisualBasic/Portable/Symbols/NamedTypeSymbol.vb]: I wonder if these can/should be cached (for `arity > 0`).

File: src/Compilers/VisualBasic/Portable/Symbols/NamedTypeSymbol.vb [hunk 8]
```diff
@@ -48,6 +48,23 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
             End Get
         End Property
 
+        ''' <summary>
+        ''' Returns custom modifiers for the type arguments that have been substituted for the type parameters. 
+        ''' </summary>
+        Friend MustOverride ReadOnly Property TypeArgumentsCustomModifiers As ImmutableArray(Of ImmutableArray(Of CustomModifier))
+
+        Friend Function CreateEmptyTypeArgumentsCustomModifiers() As ImmutableArray(Of ImmutableArray(Of CustomModifier))
```
Comment (gafter) [inline src/Compilers/VisualBasic/Portable/Symbols/NamedTypeSymbol.vb]: Shared?

Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Symbols/NamedTypeSymbol.vb]: It is using instance property, Arity.

File: src/Compilers/VisualBasic/Portable/Symbols/SubstitutedNamedType.vb [hunk 9]
```diff
@@ -663,6 +663,18 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
                 End Get
             End Property
 
+            Friend NotOverridable Overrides ReadOnly Property TypeArgumentsCustomModifiers As ImmutableArray(Of ImmutableArray(Of CustomModifier))
+                Get
+                    Return ArrayBuilder(Of ImmutableArray(Of CustomModifier)).GetInstance(Arity, ImmutableArray(Of CustomModifier).Empty).ToImmutableAndFree()
```
Comment (gafter) [inline src/Compilers/VisualBasic/Portable/Symbols/SubstitutedNamedType.vb]: Can this kind of code be centralized to we can cache such values?

File: src/Compilers/VisualBasic/Portable/Symbols/NamedTypeSymbol.vb (hunk 7 above)
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Symbols/NamedTypeSymbol.vb]: Let's wait and see what impact does this change have on perf. Compiler usually doesn't request this array if HasTypeArgumentsCustomModifiers is false.

File: src/Compilers/VisualBasic/Portable/Symbols/SubstitutedNamedType.vb (hunk 9 above)
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Symbols/SubstitutedNamedType.vb]: I will move it into a helper function, but I will not add caching until perf numbers indicate it is needed.

**Comment by gafter** [comment]
LGTM

**Comment by AlekseyTs** [comment]
@gafter, @jaredpar, @VSadov, @agocke Please review second commit, it includes support for C# and   changes addressing review feedback.

**Comment by AlekseyTs** [comment]
@dotnet-bot Test this please.

**Comment by AlekseyTs** [comment]
@dotnet-bot Test this please.

**Comment by AlekseyTs** [comment]
@dotnet-bot Test this please.

**Comment by AlekseyTs** [comment]
@gafter, @jaredpar, @VSadov, @agocke Please review second commit, it includes support for C# and changes addressing review feedback.

**Comment by gafter** [comment]
Looking now...

File: src/Compilers/CSharp/Portable/Emitter/EditAndContinue/CSharpSymbolMatcher.cs [hunk 10]
```diff
@@ -662,6 +677,9 @@ private static MethodSymbol SubstituteTypeParameters(MethodSymbol method)
             private bool AreNamedTypesEqual(NamedTypeSymbol type, NamedTypeSymbol other)
             {
                 Debug.Assert(s_nameComparer.Equals(type.Name, other.Name));
+                // TODO: Test with overloads (from PE base class?) that have modifiers.
+                Debug.Assert(!type.HasTypeArgumentsCustomModifiers);
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Emitter/EditAndContinue/CSharpSymbolMatcher.cs]: Why is this true?

Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Emitter/EditAndContinue/CSharpSymbolMatcher.cs]: For some reason this class already makes this assumption for other types that can have modifiers. For example, AreArrayTypesEqual (line 626), etc. I am going to open a work item to confirm whether these are valid assumptions and add necessary handling if not.

**Comment by jaredpar** [comment]
LGTM modulo some mostly minor comments.

**Comment by gafter** [comment]
LGTM. (@jaredpar I believe the reference versus value equality change is OK.)

**Pull request merged**